Documentation/installation/k8s-install-migration.rst

```diff
@@ -0,0 +1,310 @@
+.. only:: not (epub or latex or html)
+
+    WARNING: You are looking at unreleased Cilium documentation.
+    Please use the official rendered version released here:
+    https://docs.cilium.io
+
+.. _cni_migration:
+
+*************************************
+Migrating a cluster to Cilium
+*************************************
+
+Cilium supports "adopting" existing clusters. Running clusters can
+be migrated on a node-by-node basis, without disrupting existing traffic
+or requiring a complete cluster outage or rebuild.
+
+This document outlines how migrations with Cilium work. You will have a good
+understanding of the basic requirements, as well as see an example migration
+which you can practice using :ref:`Kind <gs_kind>`
+
+
+Background
+==========
+
+When a Pod is created, the installed CNI is called which will attach a network
+interface to the Pod. If all goes well, this network interface will join that
+Pod to the cluster network. The network interface shares the same life cycle
+as the Pod, meaning that only newly-created Pods will be picked up by the newer
+CNI if it was swapped out. This requires that all Pods on the cluster that are
+a part of the cluster network be recycled in order to be a member of the new CNI.
+
+A naive approach to migrating a CNI would be to gradually roll each node on the
+cluster, and then replacing the CNI installed when the node is brought back up.
+This is however, not possible—the second CNI will be installed with a separate
+CIDR range to that of the currently installed CNI, which has no knowledge of
+the other network range. Not to mention the use of different encapsulation
+protocols, or broken network policy since identity is lost.
+
+Instead, Cilium supports a "hybrid" mode, where two separate overlays are installed
+across the cluster. While pods on a given node can only be attached to one network,
+they have access to both Cilium and non-Cilium pods while the migration is
+taking place. As long as Cilium and the existing networking provider use a separate
+IP range, the Linux routing table takes care of separating traffic.
```

Comment:
```suggestion
When the kubelet creates a Pod, the installed CNI, as configured in /etc/cni/net.d/, is called. The cni will handle the networking for a pod. From allocating an ip address to ensuring that the pod interface is configured appriopriately as part of the pod network. The Pods network configuration shares the same life cycle
as the Pod, In the case of migration, we typically reconfigure /etc/cni/net.d/ to point to the new cni. Once completed any existing pods will have been configured by the old cni and any new pods will be configured by the newer
CNI. To complete the migration all Pods on the cluster that are
configured by the old cni must be recycled in order to be a member of the new CNI.

A naive approach to migrating a CNI would be to reconfigure all nodes with a new CNI and then gradually restart each node in the
cluster, thus replacing the CNI when the node is brought back up and ensuring that all pods are part of the new CNI.

This simple migration while effective comes at the cost of restarting all pods and nodes cluster wide before proceeding. This will be a significant outage for the cluster.

In this document we will discuss a model for live migrating between two deployed CNI implementations. This will have the benefit of reducing downtime of nodes and workloads and ensuring that workloads on both configured CNIs can communicate during migration. 

For live migration to work the second CNI will be installed with a separate
CIDR range to that of the currently installed CNI, which has no knowledge of
the other network range. Not to mention the use of different encapsulation
protocols. If the cluster is making use of network policies or other capabilities that are cni specific. This usually adds significant challenges to the migration model.

To help address the case of live migration as of 1.13 Cilium supports a "hybrid" mode.

This migration strategy allows us to configure each nodes CNI implementation independently. Allowing you to migrate nodes at your own pace and ensuring that workloads on those nodes are able to reach other workloads still deployed using the old CNI while the migration is taking place. 
 As long as Cilium and the existing CNI use a separate
IP range, the Linux routing table takes care of separating traffic.
```